Fixed file resource raising UndefinedMethod on source_path

zenspider opened the pull request:
This is a first pass at fixing the failure found by @scottvidmar where
file resource was raising undefined_method on nil. The problem was
that the initialize method for file was not being called at all
because Resource#initialize was returning early without reporting
anything.

This fix causes a failure in test/functional/inspec_exec_test.rb where
it is testing a similar scenario but for some reason the unsupported
resources get reported. We need to figure that part out, but in the
meantime this will at least report an error at the root cause rather
than down the road.

Fixes #4208

Paired with @miah 

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed file resource raising UndefinedMethod on source_path

This is a first pass at fixing the failure found by @scottvidmar where
file resource was raising undefined_method on nil. The problem was
that the initialize method for file was not being called at all
because Resource#initialize was returning early without reporting
anything.

This fix causes a failure in test/functional/inspec_exec_test.rb where
it is testing a similar scenario but for some reason the unsupported
resources get reported. We need to figure that part out, but in the
meantime this will at least report an error at the root cause rather
than down the road.

Fixes #4208

Signed-off-by: Ryan Davis <zenspider@chef.io>

jerryaldrichiii reviewed (commented):
This looks great! Is this a WIP PR and we plan on adding tests for this specific use case? (e.g. a functional test of using aws target with a linux profile)?

jerryaldrichiii reviewed (approved):
Approving the implementation as it exists now, pending tests.

Thanks for this. Not presenting the user with NoMethodError is valuable in its own right.

clintoncwolfe commented on lib/inspec/plugin/v1/plugin_types/resource.rb:
```diff
@@ -83,11 +83,14 @@ def initialize(backend, name, *args)
           test_backend = defined?(Train::Transports::Mock::Connection) && backend.backend.class == Train::Transports::Mock::Connection
-          # do not return if we are supported, or for tests
-          return unless supported || test_backend
+          # raise if we are supported, or for tests
```
I think the comment contradicts the code, here. (I think the original did too)

clintoncwolfe reviewed (approved):
A minor request to fix a comment, but otherwise fine. Thanks!

clintoncwolfe commented on lib/inspec/plugin/v1/plugin_types/resource.rb:
```diff
@@ -83,11 +83,14 @@ def initialize(backend, name, *args)
+          unless supported || test_backend
+            msg = "Unsupported resource/backend combination: %s / %s. Exiting." %
+              [name, backend.platform.name]
+            raise ArgumentError, msg
```
Here's an idea, which would at least improve messaging - create a custom exception class (or we may have a suitable one already - see lib/exceptions or lib/errors, I know we have two). Perhaps ResourceUnsupportedError? Then add a `rescue` matching that new exception class in the control evaluator - https://github.com/inspec/inspec/blob/954e6c0ff98113ed41e95b50f5a6651a30572c8c/lib/inspec/rule.rb#L54 . It make a special case of this, but would give you the chance to handle the outcome with a little more finesse.

zenspider committed:
Code cleanup.

While I was in here... nothing big.

It does seem like we have too many hard-coded symbols in our code in
general as well as too many symbol styles (underscore vs dash).

Signed-off-by: Ryan Davis <zenspider@chef.io>

miah reviewed (changes requested):
Lets pair on this, I think we should try to bubble this up through another exception. If we can't this is fine for now, but we should file a issue to revisit this exit path in the future.

zenspider committed:
test change for bad resource/backend combo.

I think this is probably in contention with my other PR on resource
initialization. We definitely need to change that, so this might fall
to the side. Or maybe just the test needs to be redone. We'll discuss
and see.

Signed-off-by: Ryan Davis <zenspider@chef.io>
